src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseDataFileService.cs

```diff
@@ -428,89 +432,155 @@ public async Task<Either<ActionResult, DataFileInfo>> UploadAsZip(Guid releaseVe
                 });
         }
 
-        public async Task<Either<ActionResult, List<DataFileInfo>>> UploadAsBulkZip(
+        public async Task<Either<ActionResult, List<ArchiveDataSetFileViewModel>>> ValidateAndUploadBulkZip(
             Guid releaseVersionId,
-            IFormFile bulkZipFormFile)
+            IFormFile zipFile,
+            CancellationToken cancellationToken)
         {
             return await _persistenceHelper
                 .CheckEntityExists<ReleaseVersion>(releaseVersionId)
                 .OnSuccess(_userService.CheckCanUpdateReleaseVersion)
-                .OnSuccess(_ => _dataArchiveValidationService.ValidateBulkDataArchiveFile(
-                        releaseVersionId, bulkZipFormFile)
-                .OnSuccess(async dataArchiveFiles =>
+                .OnSuccessVoid(async _ =>
                 {
-                    var bulkZipFile = new File
+                    var errors = await _dataArchiveValidationService.IsValidZipFile(zipFile);
+                    if (errors.Count > 0)
                     {
-                        CreatedById = _userService.GetUserId(),
-                        RootPath = releaseVersionId,
-                        ContentLength = bulkZipFormFile.Length,
-                        ContentType = bulkZipFormFile.ContentType,
-                        Filename = bulkZipFormFile.FileName.ToLower(),
-                        Type = BulkDataZip,
-                    };
-                    _contentDbContext.Files.Add(bulkZipFile);
-                    await _contentDbContext.SaveChangesAsync();
+                        return new Either<ActionResult, Unit>(Common.Validators.ValidationUtils.ValidationResult(errors));
+                    }
 
-                    // each data/meta file pair are extracted to blob storage by _dataImportService.Import
-                    await UploadFileToStorage(bulkZipFile, bulkZipFormFile);
+                    return Unit.Instance;
+                })
+                .OnSuccess(async _ => await _dataArchiveValidationService.ValidateBulkDataArchiveFiles(releaseVersionId, zipFile))
+                .OnSuccess(async dataSetFiles =>
+                {
+                    await using var stream = zipFile.OpenReadStream();
+                    var archive = new ZipArchive(stream);
 
-                        var results = new List<DataFileInfo>();
+                    var viewModels = new List<ArchiveDataSetFileViewModel>();
+
+                    foreach (var file in dataSetFiles)
+                    {
+                        var dataFileEntry = archive.GetEntry(file.DataFilename);
+                        var metaFileEntry = archive.GetEntry(file.MetaFilename);
 
-                        foreach (var archiveFile in dataArchiveFiles)
+                        await UploadFileToTempStorage(file.DataFilePath, dataFileEntry!, cancellationToken);
+                        await UploadFileToTempStorage(file.MetaFilePath, metaFileEntry!, cancellationToken);
```

Comment:
I think we can generate Guids for the data and meta file here. The path can be generated inside `UploadFileToTempStorage` by passing it extra arguments so we can do `$"{FileStoragePathUtils.FilesPath(releaseVersionId, fileType)}{fileId}}"

Then can also remove `DataFilePath` and `MetaFilePath` from `ArchiveDataSetFile` and `GetFileIdFromPath`.